Add USGS data fetch and munge

amsnyder opened the pull request.

amsnyder commented:
I added in the daily average calculation, under the assumption that we need half of daily measurements to use the value. I noticed in some of my data files the time step of the measurements changed for a given station, even within the one year of data currently being pulled (e.g. started at 30 minute measurements and had changed to 15 minute measurements). I wrote the code so that it would check how many timestamps were available for a day and calculate the proportion that actually had data. This proportion needed to be greater than 0.5, like we talked about, but we can change that any time.

galengorski commented on 01_fetch/src/fetch_usgs.py:
```diff
@@ -0,0 +1,31 @@
+
+def fetch_data(site_list, start_dt, end_dt, outfile):
+    '''fetch USGS NWIS data for locations in site_list (gets all parameters available)'''
+    for site_num in site_list:
```
Do you think putting the for loop in the function is better than looping over the function? I'm not sure, just thinking about it from a readability standpoint?

amsnyder replied:
For me, they are both equally readable. It is nice that the outfile gets formatted before putting it into the function if we pull the for loop out. I've pull it our here and in the munge function.

galengorski commented on 02_munge/src/munge_usgs.py:
```diff
@@ -0,0 +1,79 @@
+    # >     The value is known to be greater than reported value and is write protected.
+    # 1     Value is write protected without any remark code to be printed
+    # 2     Remark is write protected without any remark code to be printed
+    flags_to_drop = ['e', '&', 'E', 'P', '<', '>', '1', '2']
```
This is great, I think what you've written makes sense, but we might want to go back and take a look at when these flags pop up within the timeseries (maybe you've already done this).

amsnyder replied:
P (provisional data) was really the main flag I saw popping up a lot, and Jared suggested we not use that data. I noticed a lot of the data files Salme put on Sharepoint included provisional data, and now that data is gone when I fetch from USGS--probably a sign that provisional data is often bad.

amsnyder replied:
All of the others sounded kind of like estimations to me too, so I figured if we don't want provisional data, we probably don't want other estimates. The only exception might be for flags 1 and 2...I could try to look into why those might show up.

Are there any flags in particular that you think we might want to use data for?

galengorski replied:
Not that I can think of based on their descriptions

galengorski commented on 02_munge/src/munge_usgs.py:
```diff
@@ -40,8 +40,14 @@ def process_data_to_csv(raw_datafiles, flags_to_drop):
 
-        # (to be added in future) aggregate data to specific timestep?
-        # df = df.groupby([df['datetime'].dt.date])[cols].mean()
+        # aggregate data to daily timestep
```
Given what Salme said today, it might make sense to make our aggregation time scale (i.e. daily, hourly etc) a variable if possible

amsnyder replied:
Good point. I've created a variable to specify this. The way I set it up, we have to add a little code for each timestep we might want to process, so right now 'daily' is the only option. I can add more options as the need arises.

There might be a more sleek way to do this, but I'll dig into that when we add more timestep processing levels. For now, at least it draws attention to the fact that this is a variable we might need to change. Let me know if you think this works for now!

galengorski commented:
@amsnyder This is looking great so far, very exciting to have our first PR! I am still getting used to reviewing PRs so let me know if what I have done is reasonable. I added a few comments, but nothing major. I think that having >= 0.5 times the measurements for the time period makes sense at least for now as you said. I'll leave this for @ted80810 to look over too, but I think it looking good.

amsnyder commented:
@galengorski I think I have addressed your comments! Let me know if you have any more suggestions!

@ted80810 - could I get your review on this, and we can merge it once you both approve?

galengorski commented:
I'm not sure how to approve my review without merging the PR, and I want to make sure Ted has a chance to review it. Is there a way to do this?

amsnyder commented:
@galengorski I think you should see some kind of 'Approve' button at the bottom of this thread in the 'Review requested' section. Approving won't do anything...only once you click the 'Merge pull request' button will it merge. If that's not what you're seeing, I might need to screenshare with you.

amsnyder commented on 01_fetch/src/fetch_usgs.py:
```diff
@@ -0,0 +1,41 @@
+def main():
+    # start S3 session so that we can upload data
+    session = boto3.Session(profile_name='dev')
+    s3_client = session.client('s3')
```
```suggestion
    s3_client = session.client('s4')
```

amsnyder reviewed (commented):
check this out

ted80810 commented on 01_fetch/src/fetch_usgs.py:
```diff
@@ -0,0 +1,54 @@
+
+    site_ids = ['01411390', '01463500', '01464040', '014670261', '01467059', '01467200',
+    '01474500', '01474703', '01477050', '01482695', '01482800']
+    start_dt = '2019-01-01'
```
Do you think the fetch_data function should be adjusted to accept different date formats (ie. 20191231) from users or specify the input date format in a comment? I ask because the NOAA NOS url has a slightly different format, but I want to know your thoughts.

amsnyder replied:
I just added a comment to suggest the date format - let me know if you think that works for now! I think we can have different formats in the different scripts for now, but we will revisit if it becomes a problem once we start to build the data pipeline out more!

ted80810 replied:
Yea! that works. Sounds good to me.

amsnyder commented on 01_fetch/src/fetch_usgs.py:
```diff
@@ -0,0 +1,54 @@
+
+    site_ids = ['01411390', '01463500', '01464040', '014670261', '01467059', '01467200',
+    '01474500', '01474703', '01477050', '01482695', '01482800']
+    start_dt = '2019-01-01'
```
```suggestion
    # start and end dates should be in the format 'YYYY-MM-DD'
    start_dt = '2019-01-01'
```

ted80810 commented:
Everything looks great! Super excited about implementing the S3 storage. Thank you spending your time putting these scripts together.
